pkg/virt-controller/watch/vm.go

```diff
@@ -2899,6 +2929,20 @@ func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
 	vm.Status.VolumeRequests = tmpVolRequests
 }
 
+func liveUpdateVolumes(lastSeenVMSpec *virtv1.VirtualMachineSpec, vm *virtv1.VirtualMachine) {
+	oldVols := make(map[string]bool)
+	for _, v := range lastSeenVMSpec.Template.Spec.Volumes {
+		if v.DataVolume != nil || v.PersistentVolumeClaim != nil {
+			oldVols[v.Name] = true
+		}
+	}
+	for i, v := range vm.Spec.Template.Spec.Volumes {
+		if _, ok := oldVols[v.Name]; ok {
+			lastSeenVMSpec.Template.Spec.Volumes[i] = v
+		}
+	}
```

Comment:
Doesn't look like this second loop is actually doing anything.  Not changing `lastSeenVMSpec.Template.Spec.Volumes` at all.  What is the intention?

Reply:
I get it now

Reply:
I'm confused again, what about this assignment? https://github.com/kubevirt/kubevirt/blob/7b5c9e54a8afd8c19b25e8afb611470b66294c77/pkg/virt-controller/watch/vm.go#L2955

`lastSeenVMSpec.Template.Spec.Volumes` should equal `vm.Spec.Template.Spec.Volumes`


What is the intention of this function?

Reply:
So this follows the [same mechanism](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/vm.go#L2910-L2921) as the other live-updateable fileds.

The first loop saves the kind of volumes that can be live-upated, PVCs and DVs (for example we cannot live-update container disks). The second loop updates the volumes in the lastSeenVMSpec, in this way the [DeepEqual](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/vm.go#L2923) is true and the generic condition isn't applied

Reply:
Example with live-updateable volumes:
- lastSeenVMSpec volumes: {Name: vol1, &DV{Name:"dv-src"}}
- vm volumes: {Name: vol1, &PVC{Name:"pvc-dst"}}
-> result:  lastSeenVMSpec volumes: {Name: vol1, &PVC{Name:"pvc-dst"}} -> DeepEqual: true && generic restart condition not set

Example with no updateable volumes:
- lastSeenVMSpec volumes: {Name: vol1, &ContainerDisk{Name:"cd"}}
- vm volumes: {Name: vol1, &PVC{Name:"pvc-dst"}}
-> result:  lastSeenVMSpec volumes: {Name: vol1, &ContainerDisk{Name:"cd"}} -> DeepEqual: false && generic restart condition set

Reply:
> lastSeenVMSpec volumes: {Name: vol1, &DV{Name:"dv-src"}}
   vm volumes: {Name: vol1, &PVC{Name:"pvc-dst"}}

In your example `lastSeenVMSpec` and `vm` are different but I don't see how that is possible given this assignment:

https://github.com/kubevirt/kubevirt/blob/7b5c9e54a8afd8c19b25e8afb611470b66294c77/pkg/virt-controller/watch/vm.go#L2955

Reply:
Ah good catch!
Mmh.. not quite sure why we also consider volumes there?

Reply:
This has been discussed and fixed in https://github.com/kubevirt/kubevirt/pull/11770